clustered/common/src/main/java/org/ehcache/clustered/common/internal/messages/ResponseCodec.java

```diff
@@ -39,126 +94,143 @@
+          .encode().array();
       }
       case SERVER_INVALIDATE_HASH: {
         ServerInvalidateHash serverInvalidateHash = (ServerInvalidateHash) response;
-        buffer = ByteBuffer.allocate(OP_CODE_SIZE + serverInvalidateHash.getCacheId().length() * 2 + 8);
-        buffer.put(EhcacheEntityResponse.Type.SERVER_INVALIDATE_HASH.getOpCode());
-        CodecUtil.putStringAsCharArray(buffer, serverInvalidateHash.getCacheId());
-        buffer.putLong(serverInvalidateHash.getKey());
-        return buffer.array();
+        return SERVER_INVALIDATE_HASH_RESPONSE_STRUCT.encoder()
+          .enm(RESPONSE_TYPE_FIELD_NAME, serverInvalidateHash.getResponseType())
+          .string(SERVER_STORE_NAME_FIELD, serverInvalidateHash.getCacheId())
+          .int64(KEY_FIELD, serverInvalidateHash.getKey())
+          .encode().array();
       }
       case MAP_VALUE: {
         MapValue mapValue = (MapValue) response;
         byte[] encodedMapValue = Util.marshall(mapValue.getValue());
-        buffer = ByteBuffer.allocate(OP_CODE_SIZE + encodedMapValue.length);
-        buffer.put(EhcacheEntityResponse.Type.MAP_VALUE.getOpCode());
-        buffer.put(encodedMapValue);
-        return buffer.array();
+        return MAP_VALUE_RESPONSE_STRUCT.encoder()
+          .enm(RESPONSE_TYPE_FIELD_NAME, mapValue.getResponseType())
+          .byteBuffer("mapValue", wrap(encodedMapValue))
+          .encode().array();
       }
       default:
-        throw new UnsupportedOperationException("The operation is not supported : " + response.getType());
+        throw new UnsupportedOperationException("The operation is not supported : " + response.getResponseType());
     }
   }
 
   public EhcacheEntityResponse decode(byte[] payload) {
-    ByteBuffer buffer = ByteBuffer.wrap(payload);
-    byte opCode = buffer.get();
-    EhcacheEntityResponse.Type type = EhcacheEntityResponse.Type.responseType(opCode);
-    byte[] payArr = new byte[buffer.remaining()];
-    buffer.get(payArr);
-    switch (type) {
+    ByteBuffer buffer = wrap(payload);
+    StructDecoder decoder = SUCCESS_RESPONSE_STRUCT.decoder(buffer);
+    Enm<EhcacheResponseType> opCodeEnm = decoder.enm(RESPONSE_TYPE_FIELD_NAME);
+
+    if (!opCodeEnm.isFound()) {
+      throw new AssertionError("Got a response without an opCode");
+    }
+    if (!opCodeEnm.isValid()) {
+      // Need to ignore the response here as we do not understand its type - coming from the future?
+      return null;
+    }
+
+    EhcacheResponseType opCode = opCodeEnm.get();
+    buffer.rewind();
+    switch (opCode) {
       case SUCCESS:
         return EhcacheEntityResponse.Success.INSTANCE;
       case FAILURE:
-        ClusterException exception = (ClusterException)Util.unmarshall(payArr);
+        decoder = FAILURE_RESPONSE_STRUCT.decoder(buffer);
+        ClusterException exception = (ClusterException)Util.unmarshall(decoder.byteBuffer(EXCEPTION_FIELD));
```

Comment:
Don't you want to also (un)marshall exceptions with runnel? Otherwise you're running the classic risk of having problems deserializing the exception between two different versions of a client and a server.